src/pipelines/epidemiology/config.yaml

```diff
@@ -117,7 +117,7 @@ sources:
       country: AT
       sep: ";"
     test:
-      location_key_match: 'AT_.+'
+      location_key_match: 'AT_[^_]+[^_]+'
```

Comment:
Hmm, I thought I commented on this already but don't see it.  When I make this change, 

update.py --only epidemiology --location-key AT_*_*
and
update.py --only epidemiology --location-key AT_*

both only run the level 2 data source.

Reply:
That sounds possible, I think the logic which filters data sources from the location key parameter is still pretty buggy.

Reply:
location_key_match seems to serve two purposes.

It is used in the data source test to match the metadata key in https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/18dbd883b1d18c3d53a8c524aee2154934bea563/src/test/test_source_run.py

And it is used in update.py to filter which pipelines to run, by matching the location-key parameter.

location_key_match: 'AT_[^_]+[^_]+' seems to be fine for matching the metadata key.

However, because the metadata keys for level 2 are of the form 'AT' or 'AT_' followed by a number, and the keys for level 3 are of the form 'AT_' followed by three numbers, any location_key_match that matches every level 2 string (note that the existing 'AT_' does not capture the country-level key 'AT'; is that a problem in itself?) would also match every level 3 string.

Does location_key_match need to capture every metadata key that the source matches?  If so, it's not possible to set up level 2 and level 3 location_key_match strings that allow running only the level 3 source but not the level 2 source.

Perhaps it's okay though to always run the level 2 source along with the level 3 source -- if so, the existing location_key_match is fine; to run the level 3 source, we just need to specify update.py --only epidemiology --location-key AT_** or something else of the sort (since 'AT_*_*' has two underscores, it doesn't match 'AT_[^_]+[^_]+').

Reply:
> location_key_match seems to serve two purposes.

Yes, it's kind of a hack. I was trying to implement an easy way to test newly added data sources without having to comment out all other data sources from the `config.yaml` file.

> Does location_key_match need to capture every metadata key that the source matches? If so, it's not possible to set up level 2 and level 3 location_key_match strings that allow running only the level 3 source but not the level 2 source.

Right, `AT_*` should indeed match levels 2 and 3. It's not even proper RegEx, it should be `AT_.+`. If you can think of a better way to make this work, I'm open to it. Originally I was thinking something like `update.py [...] --country XX --level [0,1,2,3]` and then we can build the proper RegEx automatically without having to type it all out on the command line. Alternatively we can process the `location_key` parameter as a non-regex string and replace `*` with `[^_]+`.

> since 'AT__' has two underscores,

Which location has two underscores? That sounds like a bug.

Reply:
No location has two underscores.  I was using raw text and it removed the star; I meant to say `AT_*_*`,
referring to the location-key parameter you had directed me to use for testing level 3 by itself.

Building the proper RegEx string for a country and level will depend on the format of that country's keys, so I don't see how this can be generalized.  Instead of location_key_match, could we have update.py take in a class name, list of class names, or regex of class names to run (such as pipelines.epidemiology.at_authority.AustriaLevel3DataSource) ?

Reply:
Filtering by class name is a great idea, and it would be a good addition. I don't think it solves the use case I had in mind, which is: what is the data for location `X` after these changes? Adding all the classes by hand sounds very error prone.

Say that we parsed each `*` as `[^_].+` and sandwiched the expression between `^` and `$`. Then `AT_*` would become `^AT_[^_]+$`, which should only match level 1. And `AT_*_*` would become `^AT_[^_]+_[^_]+$` which should only match level 2.

Even though this would depend on data sources providing the right regex, we can enforce that via unit tests. Currently we check if data sources output at least one row, we can easily add [another test](https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/main/src/test/test_source_run.py#L89) to check that all output location keys match what's in the test string. Keep in mind that data sources can provide a list of regex, not just a single value ([example](https://github.com/GoogleCloudPlatform/covid-19-open-data/blob/main/src/pipelines/demographics/config.yaml#L109)).

If we get this to work well, we can re-add the old functionality which told us what outputs changed with each PR, since running a small subset of data sources would be much more feasible. For example, imagine adding a tag to the PR description like `loc:AT_*_*` which would trigger a data pipeline run using `--location-key AT_*_*` and attach the output to the PR discussion...

Reply:
I guess if we're going to replace `*` with `[^_]+` we could also do it for the test parameter in the `config.yaml`, since using plain regex is very powerful but also extremely error prone...

Reply:
I think I was mistaken -- I've been referring to the country as level 1, state as 2 and district as 3, but I see here that state is 1 and district is 2 (so I suppose we're probably using 0 to denote country?): [https://screenshot.googleplex.com/5HwZobRsSdskyvg](https://screenshot.googleplex.com/5HwZobRsSdskyvg)

`^AT_[^_]+$` matches levels 1 and 2 but not the country level:


```
(base) mgl@mgl-macbookpro:~/src/covid-19-open-data/src$ python
Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import re
>>> re.match('^AT_[^_]+$', 'AT')
>>> re.match('^AT_[^_]+$', 'AT_1')
<re.Match object; span=(0, 4), match='AT_1'>
>>> re.match('^AT_[^_]+$', 'AT_101')
<re.Match object; span=(0, 6), match='AT_101'>

```

But good idea to sandwich the expression between `^` and `$`, and thanks for letting me know we can use a list of regex.  I don't think we need to parse `*` as `[^_].+`.  What if we just used the list [`AT`, `AT_.`] for the level 0/1 source and used `AT_...` for the level 2 source?

```
>>> re.match('^AT$', 'AT')
<re.Match object; span=(0, 2), match='AT'>
>>> re.match('^AT_.$', 'AT_1')
<re.Match object; span=(0, 4), match='AT_1'>
>>> re.match('^AT$', 'AT_101')
>>> re.match('^AT_.$', 'AT_101')
>>> re.match('^AT_...$', 'AT_101')
<re.Match object; span=(0, 6), match='AT_101'>

```

Reply:
> I think I was mistaken -- I've been referring to the country as level 1, state as 2 and district as 3, but I see here that state is 1 and district is 2 (so I suppose we're probably using 0 to denote country?)

Yes, we generally use the notation country level, admin level 1 (L1) and admin level 2 (L2). I think some of the comments and class names are not consistent with this...

> But good idea to sandwich the expression between ^ and $, and thanks for letting me know we can use a list of regex. I don't think we need to parse * as [^_].+. What if we just used the list [AT, AT_.] for the level 0/1 source and used AT_... for the level 2 source?

Your comment made me realize that the keys for AT admin level 2 are not correct. They need to be in the form `AT_${subregion1_code}_${subregion2_code}` (admin level 2 should *always* have two underscores). So, if the `subregion2_code` is `108`, then the key needs to be `AT_1_108`. If you think the `108` is unnecessary (since `1` should be obvious from `subregion1_code`) then you can change the `subregion2_code` to just the last 2 digits (i.e. key `AT_1_08`). I'll leave it up to you.

Reply:
Changed the numbering of the administrative levels and fixed the level 2 keys.

Regarding the use case of "what is the data for location X after these changes", the current location_key_match still doesn't solve this use case for any regex X, and for example wouldn't be able to produce the data for Austria level 1 only but not the country-level, since they're both in the same source.  I suppose this isn't critical, though.

I'm hesitant to automatically replace `*` with `[^_]+` (and more weakly object to padding with `^` and `$`) because it creates a modified regex notation that can be confusing.  If keys are standard enough to support it, though, I do like the idea of  something like `update.py [...] --country XX --level [0,1,2,3]` and building the proper RegEx automatically without having to type it all out on the command line.  This still doesn't solve the problem of data sources providing data for multiple levels, though.

Reply:
> Changed the numbering of the administrative levels and fixed the level 2 keys.

LGTM, you only forgot to update the keys in one of the auxiliary tables.

> I do like the idea of something like update.py [...] --country XX --level [0,1,2,3] and building the proper RegEx automatically without having to type it all out on the command line

Fair enough. If we go with that, we can do post-filtering of the data source like we are doing when the option `--strict-match` is passed.